File: ruby_tdd/20180616_tennis/lib/tennis.rb
class Tennis
  SCORE = {
    0 => 'love',
    1 => 15,
    2 => 30,
    3 => 40,
  }

  def initialize
    @player_scores = {1 => 0, 2 => 0}
  end

  def score
    have_winner ||
      have_advantage ||
      have_deuce ||
      standard_score
  end
  def player_scores(player_id)
    @player_scores[player_id] += 1
  end

  private

  def leading_player
    @player_scores.key(@player_scores.values.max)
  end

  def player_with_winning_score
    @player_scores[leading_player] > 3
  end

  def leading_by_two_or_more
    (@player_scores.values.max - @player_scores.values.min ) >= 2
  end

  def leading_by_one
    (@player_scores.values.max - @player_scores.values.min ) >= 1
  end

  def deuce_scores
    scores_equal && @player_scores.values.first >= 3
  end

  def scores_equal
    @player_scores.values.uniq.count == 1
  end

  def have_winner
    "game player #{leading_player}" if player_with_winning_score && leading_by_two_or_more
  end

  def have_advantage
    "advantage player #{leading_player}" if player_with_winning_score && leading_by_one
  end

  def have_deuce
    "deuce" if deuce_scores
  end

  def standard_score
    scores = [1, 2].map{|player| SCORE[@player_scores[player]]}
    scores = [scores.first, 'all'] if scores_equal
    scores.join(" ")
  end

end

File: es6_tdd/20180616_tennis/tennis.js
const scores = {
  0: 'love',
  1: 15,
  2: 30,
  3: 40,
}

class Tennis {
  constructor() {
    this.playerScores = {
      1: 0,
      2: 0,
    }
  }

  scores() {
    return this.vals().map( x => scores[x] )
  }

  score() {
    return this.winnerScore()
      || this.advantageScore()
      || this.deuceScore()
      || this.standardScore()
  }

  winnerScore() {
    if (this.haveWinner()) {
      return `game player ${this.leadingPlayer()}`
    }
  }

  advantageScore() {
    if (this.secondMax() >= 3 && !this.evenScores()) {
      return `advantage player ${this.leadingPlayer()}`
    }
  }

  deuceScore() {
    if (this.evenScores() && this.min() >= 3) {
      return 'deuce'
    }
  }

  standardScore() {
    let scores = this.scores()
    if (this.evenScores()) {
      scores = Array.from(new Set(this.scores()))
      scores = scores.concat(['all'])
    }
    return scores.join(" ")
  }

  haveWinner() {
    return !this.evenScores()
      && (this.max() > 3)
      && ((this.max() - this.secondMax()) >= 2)
  }

  leadingPlayer() {
    const max = this.max()
    return this.keys().find( x => this.playerScores[x] === max )
  }

  evenScores() {
    return this.min() === this.max()
  }

  scoreByPlayer(player) {
    if (!this.playerScores[player]) this.playerScores[player] = 0
    this.playerScores[player]++
  }

  keys() {
    return Object.keys(this.playerScores)
  }

  vals() {
    return this.keys().map( x => this.playerScores[x] )
  }

  min() {
    return Math.min.apply(null, this.vals())
  }

  secondMax() {
    const valsNoMax = this.vals().filter( e => e !== this.max() )
    //console.log(this.vals())
    //console.log(valsNoMax)
    return Math.max.apply(null, valsNoMax)
  }

  max() {
    return Math.max.apply(null, this.vals())
    return 'love all'
  }
}

export default Tennis
